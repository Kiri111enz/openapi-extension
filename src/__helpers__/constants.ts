const MDSections = {
    OK: '200 OK',
    NOT_FOUND: '404 Not Found',
    TITLE: 'TEST ENDPOINT',
    REQUEST: 'Request',
};

export {MDSections};
export default {MDSections};
