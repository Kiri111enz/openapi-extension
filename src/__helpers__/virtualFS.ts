type VirtualFS = {
    mkdir(targer: string, options: any): void;
    writeFile(path: string, content: string): void;
    readFile(path: string): string;
    match(path: string): string;
    reset(): void;
    get pages(): Record<string, string>;
};

function virtualFS(): VirtualFS {
    /** virtual fs record with 1 depth */
    let pages: Record<string, string> = {};

    const fs: VirtualFS = {
        mkdir() {},
        readFile: jest.fn((path) => {
            console.log(path);
            return pages[path];
        }),
        writeFile: jest.fn((path, content) => {
            pages[path] = content;
        }),
        match(target) {
            const paths = Object.keys(pages);
            const closest = paths.find((path) => path.includes(target));

            if (!closest) {
                throw new Error(`There is not page with path: ${target}.\nPages: ${Object.keys(pages).join('\n')}`);
            }

            return pages[closest];
        },
        reset() {
            pages = {};
        },
        get pages() {
            return pages;
        },
    };

    return fs;
}


const fs = virtualFS();

export {fs};
export default {fs};
