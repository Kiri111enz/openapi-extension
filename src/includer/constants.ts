import {JSONSchema6} from 'json-schema';

export const EOL = '\n';
export const TAG_NAMES_FIELD = 'x-navtitle';
export const BLOCK = EOL.repeat(2);
export const INFO_TAB_NAME = 'Info';
export const SANDBOX_TAB_NAME = 'Sandbox';
export const CONTACTS_SECTION_NAME = 'Contacts';
export const TAGS_SECTION_NAME = 'Sections';
export const ENDPOINTS_SECTION_NAME = 'Endpoints';
export const REQUEST_SECTION_NAME = 'Request';
export const PATH_PARAMETERS_SECTION_NAME = 'Path parameters';
export const HEADERS_SECTION_NAME = 'Headers';
export const QUERY_PARAMETERS_SECTION_NAME = 'Query parameters';
export const COOKIES_SECTION_NAME = 'Cookies';
export const RESPONSES_SECTION_NAME = 'Responses';
export const HTML_COMMENTS_OPEN_DIRECTIVE = '<!--';
export const HTML_COMMENTS_CLOSE_DIRECTIVE = '-->';
export const DISABLE_LINTER_DIRECTIVE = 'markdownlint-disable-file';
export const SPEC_SECTION_NAME = 'Specification';
export const SPEC_SECTION_TYPE = 'Open API';
export const LEADING_PAGE_NAME_DEFAULT = 'Overview';

export const PRIMITIVE_JSON6_SCHEMA_TYPES =
    new Set<JSONSchema6['type']>(['string', 'boolean', 'null', 'number', 'integer']);
