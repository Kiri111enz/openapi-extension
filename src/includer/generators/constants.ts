import block from 'bem-cn-lite';

export const openapiBlock = block('openapi');
