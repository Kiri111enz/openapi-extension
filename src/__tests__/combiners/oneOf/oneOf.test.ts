import {MDSections} from '../../__helpers__/constants';
import {fs} from '../../__helpers__/virtualFS';
import {runPreset} from '../../run';

describe('oneOf operator', () => {
    beforeEach(async () => {
        fs.reset();
    });

    it('renders oneOf with no parameters in request body', async () => {
        await runPreset(__dirname);

        const sections = fs.sections('empty.md');

        expect(sections[MDSections.REQUEST]).toMatchSnapshot();
    });

    it('renders oneOf with no parameters in response body', async () => {
        await runPreset(__dirname);

        const sections = fs.sections('empty.md');

        expect(sections[MDSections.OK]).toMatchSnapshot();
    });

    it('renders oneOf with other parameters in request body', async () => {
        await runPreset(__dirname);

        const sections = fs.sections('filled.md');

        expect(sections[MDSections.REQUEST]).toMatchSnapshot();
    });

    it('renders oneOf as parameter', async () => {
        await runPreset(__dirname);

        const sections = fs.sections('parameters.md');

        expect(sections[MDSections.REQUEST]).toMatchSnapshot();
    });

    it('renders oneOf in parameters and response', async () => {
        await runPreset(__dirname);

        const sections = fs.sections('parameters.md');

        expect(sections[MDSections.OK]).toMatchSnapshot();
    });

    it('renders each table only one time', async () => {
        await runPreset(__dirname);

        const page = fs.match('parameters.md');

        expect(page).toMatchSnapshot();
    });
});
