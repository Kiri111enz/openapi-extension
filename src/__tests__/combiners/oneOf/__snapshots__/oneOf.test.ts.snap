// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oneOf operator renders each table only one time 1`] = `
"<div class="openapi">

# TEST ENDPOINT

Defines a simple get operation with no inputs and a complex

## Request

#||| **method** | **url** | **description** ||

|| 
\`\`\`
POST
\`\`\`
 | 
\`\`\`
http://localhost:8080/parameters
\`\`\`
 | 
\`\`\`
Generated server url
\`\`\`
 |||#

#### Body

{% cut "application/json" %}


\`\`\`
{
    "pet": {}
}
\`\`\`


{% endcut %}


#||| **Name** | **Type** | **Description** ||

|| pet | object | [Dog](#dog)
or [Cat](#cat) |||#

### Dog

Dog class

#||| **Name** | **Type** | **Description** ||

|| type | string |  ||

|| foo | string |  |||#

### Cat

Cat class

#||| **Name** | **Type** | **Description** ||

|| type | string |  ||

|| foo | string |  |||#

## Responses

## 200 OK

Base 200 response

#### Body

{% cut "application/json" %}


\`\`\`
{
    "pet": {}
}
\`\`\`


{% endcut %}


#||| **Name** | **Type** | **Description** ||

|| pet | object | [Dog](#dog)
or [Cat](#cat) ||

|| ...rest | oneOf | [Dog](#dog)
or [Cat](#cat) |||#
#### Or value from:

<!-- markdownlint-disable-file -->

</div>"
`;

exports[`oneOf operator renders oneOf as parameter 1`] = `
[
  "## Request",
  "#||| **method** | **url** | **description** ||",
  "|| ",
  "\`\`\`",
  "POST",
  "\`\`\`",
  " | ",
  "\`\`\`",
  "http://localhost:8080/parameters",
  "\`\`\`",
  " | ",
  "\`\`\`",
  "Generated server url",
  "\`\`\`",
  " |||#",
  "#### Body",
  "{% cut "application/json" %}",
  "\`\`\`",
  "{",
  "    "pet": {}",
  "}",
  "\`\`\`",
  "{% endcut %}",
  "#||| **Name** | **Type** | **Description** ||",
  "|| pet | object | [Dog](#dog)",
  "or [Cat](#cat) |||#",
  "### Dog",
  "Dog class",
  "#||| **Name** | **Type** | **Description** ||",
  "|| type | string |  ||",
  "|| foo | string |  |||#",
  "### Cat",
  "Cat class",
  "#||| **Name** | **Type** | **Description** ||",
  "|| type | string |  ||",
  "|| foo | string |  |||#",
]
`;

exports[`oneOf operator renders oneOf in parameters and response 1`] = `
[
  "## 200 OK",
  "Base 200 response",
  "#### Body",
  "{% cut "application/json" %}",
  "\`\`\`",
  "{",
  "    "pet": {}",
  "}",
  "\`\`\`",
  "{% endcut %}",
  "#||| **Name** | **Type** | **Description** ||",
  "|| pet | object | [Dog](#dog)",
  "or [Cat](#cat) ||",
  "|| ...rest | oneOf | [Dog](#dog)",
  "or [Cat](#cat) |||#",
  "#### Or value from:",
]
`;

exports[`oneOf operator renders oneOf with no parameters in request body 1`] = `
[
  "## Request",
  "#||| **method** | **url** | **description** ||",
  "|| ",
  "\`\`\`",
  "POST",
  "\`\`\`",
  " | ",
  "\`\`\`",
  "http://localhost:8080/empty",
  "\`\`\`",
  " | ",
  "\`\`\`",
  "Generated server url",
  "\`\`\`",
  " |||#",
  "#### Body",
  "{% cut "application/json" %}",
  "\`\`\`",
  "{}",
  "\`\`\`",
  "{% endcut %}",
  "#||| **Name** | **Type** | **Description** ||",
  "|| ...rest | oneOf | [Dog](#dog)",
  "or [Cat](#cat) |||#",
  "#### Or value from:",
  "### Dog",
  "Dog class",
  "#||| **Name** | **Type** | **Description** ||",
  "|| type | string |  ||",
  "|| foo | string |  |||#",
  "### Cat",
  "Cat class",
  "#||| **Name** | **Type** | **Description** ||",
  "|| type | string |  ||",
  "|| foo | string |  |||#",
]
`;

exports[`oneOf operator renders oneOf with no parameters in response body 1`] = `
[
  "## 200 OK",
  "Base 200 response",
  "#### Body",
  "{% cut "application/json" %}",
  "\`\`\`",
  "{}",
  "\`\`\`",
  "{% endcut %}",
  "#||| **Name** | **Type** | **Description** ||",
  "|| ...rest | oneOf | [Dog](#dog)",
  "or [Cat](#cat) |||#",
  "#### Or value from:",
]
`;

exports[`oneOf operator renders oneOf with other parameters in request body 1`] = `
[
  "## Request",
  "#||| **method** | **url** | **description** ||",
  "|| ",
  "\`\`\`",
  "POST",
  "\`\`\`",
  " | ",
  "\`\`\`",
  "http://localhost:8080/filled",
  "\`\`\`",
  " | ",
  "\`\`\`",
  "Generated server url",
  "\`\`\`",
  " |||#",
  "#### Body",
  "{% cut "application/json" %}",
  "\`\`\`",
  "{",
  "    "name": "string",",
  "    "age": 0",
  "}",
  "\`\`\`",
  "{% endcut %}",
  "#||| **Name** | **Type** | **Description** ||",
  "|| name | string |  ||",
  "|| age | number |  ||",
  "|| ...rest | oneOf | [Dog](#dog)",
  "or [Cat](#cat) |||#",
  "#### Or value from:",
  "### Dog",
  "Dog class",
  "#||| **Name** | **Type** | **Description** ||",
  "|| type | string |  ||",
  "|| foo | string |  |||#",
  "### Cat",
  "Cat class",
  "#||| **Name** | **Type** | **Description** ||",
  "|| type | string |  ||",
  "|| foo | string |  |||#",
]
`;
